File: ts/test/strictness.test.ts
/* Copyright (c) 2025 Richard Rodger and other contributors, MIT License */

// strictness.test.ts — the LENIENCY probe.
//
// @tabnas/zon is a JSONIC plugin. `new Tabnas().use(Zon)` alone cannot parse
// even `.{ .a = 1 }`, so unlike (say) the json5 plugin there is no stricter
// standalone mode to fall back to: @tabnas/jsonic's relaxed number lexer and
// its "first value wins, ignore the rest" top-level behaviour are inherited
// unconditionally by the documented stack. This suite measures how much of
// that leaks through as accepted-but-not-ZON.
//
// The INPUTS are locally authored (`test/strictness/inputs.txt`, tracked).
// The VERDICTS are not: `scripts/fetch-zigzon.sh` hands every input to the
// pinned Zig reference implementation and writes `test/strictness/cases.json`
// (generated, never committed). This repo never decides what ZON means.
//
// Same rules as zigzon.test.ts: never skips, checks both halves, compares
// values rather than "it did not throw", and is RED on purpose.

import { describe, test, after } from 'node:test'
import assert from 'node:assert'
import { existsSync, readFileSync } from 'node:fs'
import { join } from 'node:path'

import { Tabnas } from '@tabnas/parser'
import { jsonic } from '@tabnas/jsonic'
import { Zon } from '../dist/zon'

const REPO = join(__dirname, '..', '..')
const CASES = join(REPO, 'test', 'strictness', 'cases.json')

const MISSING = `
strictness probe verdicts not found at:
  ${CASES}

Generate them (idempotent):
  bash scripts/fetch-zigzon.sh

This suite deliberately FAILS rather than skipping.
`

type ProbeCase = {
  source: string
  valid: boolean
  value?: unknown
  error?: string
}

// Identical to zigzon.test.ts — see the note there.
const OPTS = { enumTag: '$enum', charAsNumber: true }

function canon(v: unknown): unknown {
  if ('number' === typeof v) {
    if (Number.isNaN(v)) return '@nan'
    if (Infinity === v) return '@inf'
    if (-Infinity === v) return '@-inf'
    return v
  }
  if (null === v || 'object' !== typeof v) return v ?? null
  if (Array.isArray(v)) return v.map(canon)
  const out: Record<string, unknown> = {}
  for (const k of Object.keys(v as object)) {
    out[k] = canon((v as Record<string, unknown>)[k])
  }
  return out
}

function label(src: string): string {
  const one = src.replace(/\n/g, '\\n').replace(/\t/g, '\\t')
  return 60 < one.length ? one.slice(0, 57) + '...' : one
}

describe('strictness probe (verdicts from zig 0.16.0)', () => {
  if (!existsSync(CASES)) {
    test('probe verdicts present', () => assert.fail(MISSING))
    return
  }

  const doc = JSON.parse(readFileSync(CASES, 'utf8'))
  const cases: ProbeCase[] = doc.cases
  assert.ok(Array.isArray(cases) && 0 < cases.length, 'probe has no cases')

  const valid = cases.filter((c) => c.valid)
  const invalid = cases.filter((c) => !c.valid)
  assert.ok(0 < valid.length, 'probe has no valid inputs')
  assert.ok(0 < invalid.length, 'probe has no invalid inputs')

  let validOk = 0
  let invalidOk = 0

  describe('zig-valid inputs parse to the reference value', () => {
    for (const c of valid) {
      test(label(c.source), () => {
        const tn = new Tabnas().use(jsonic).use(Zon, OPTS)
        assert.deepStrictEqual(canon(tn.parse(c.source)), c.value)
        validOk++
      })
    }
  })

  describe('zig-invalid inputs are rejected', () => {
    for (const c of invalid) {
      test(label(c.source), () => {
        const tn = new Tabnas().use(jsonic).use(Zon, OPTS)
        let threw = false
        let got: unknown
        try {
          got = tn.parse(c.source)
        } catch {
          threw = true
        }
        assert.ok(
          threw,
          `LENIENCY LEAK: accepted as ${JSON.stringify(got)} ` +
          `an input zig rejects (${c.error})`,
        )
        invalidOk++
      })
    }
  })

  after(() => {
    const pct = (a: number, b: number) => (100 * a / b).toFixed(1) + '%'
    process.stderr.write(
      `\n[strictness] TS baseline @ zig ${doc.source.zig}\n` +
      `[strictness]   valid-accepted   ${validOk}/${valid.length} (${pct(validOk, valid.length)})\n` +
      `[strictness]   invalid-rejected ${invalidOk}/${invalid.length} (${pct(invalidOk, invalid.length)})\n`,
    )
  })
})

File: go/strictness_test.go
// Copyright (c) 2025 Richard Rodger and other contributors, MIT License

package tabnaszon

// strictness_test.go — the LENIENCY probe, Go half.
//
// @tabnas/zon is a JSONIC plugin: the Zon plugin alone cannot parse even
// `.{ .a = 1 }`, so there is no stricter standalone mode to fall back to and
// jsonic's relaxed lexer is inherited unconditionally. This suite measures how
// much of that leaks through as accepted-but-not-ZON.
//
// Inputs are locally authored (test/strictness/inputs.txt, tracked); verdicts
// come from the pinned Zig reference implementation via
// scripts/fetch-zigzon.sh (test/strictness/cases.json, generated, never
// committed).
//
// Runs the SAME inputs, verdicts and options as ts/test/strictness.test.ts.
// Never skips. RED on purpose.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

type probeCase struct {
	Source string          `json:"source"`
	Valid  bool            `json:"valid"`
	Value  json.RawMessage `json:"value"`
	Error  string          `json:"error"`
}

type probeCorpus struct {
	Source struct {
		Zig    string `json:"zig"`
		Commit string `json:"commit"`
	} `json:"source"`
	Cases []probeCase `json:"cases"`
}

const probeMissing = `
strictness probe verdicts not found at: %s

Generate them (idempotent):
  bash scripts/fetch-zigzon.sh

This suite deliberately FAILS rather than skipping.
`

func probeLabel(src string) string {
	one := strings.ReplaceAll(strings.ReplaceAll(src, "\n", `\n`), "\t", `\t`)
	if 60 < len(one) {
		one = one[:57] + "..."
	}
	return one
}

func TestStrictnessProbe(t *testing.T) {
	path := filepath.Join("..", "test", "strictness", "cases.json")
	body, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf(probeMissing, path)
	}

	var corpus probeCorpus
	if err := json.Unmarshal(body, &corpus); err != nil {
		t.Fatalf("probe %s is not valid JSON: %v", path, err)
	}
	if len(corpus.Cases) == 0 {
		t.Fatalf("probe %s has no cases", path)
	}

	var valid, invalid []probeCase
	for _, c := range corpus.Cases {
		if c.Valid {
			valid = append(valid, c)
		} else {
			invalid = append(invalid, c)
		}
	}
	if len(valid) == 0 {
		t.Fatalf("probe has no valid inputs")
	}
	if len(invalid) == 0 {
		t.Fatalf("probe has no invalid inputs")
	}

	validOk, invalidOk := 0, 0

	t.Run("zig-valid inputs parse to the reference value", func(t *testing.T) {
		for _, c := range valid {
			t.Run(probeLabel(c.Source), func(t *testing.T) {
				got, err := zigParse(c.Source)
				if err != nil {
					t.Fatalf("rejected an input zig accepts: %v", err)
				}
				want := zigWant(t, c.Value)
				if g := zigCanon(got); !reflect.DeepEqual(g, want) {
					t.Fatalf("value mismatch\n  got  %#v\n  want %#v", g, want)
				}
				validOk++
			})
		}
	})

	t.Run("zig-invalid inputs are rejected", func(t *testing.T) {
		for _, c := range invalid {
			t.Run(probeLabel(c.Source), func(t *testing.T) {
				got, err := zigParse(c.Source)
				if err == nil {
					t.Fatalf("LENIENCY LEAK: accepted as %#v an input zig rejects (%s)",
						zigCanon(got), c.Error)
				}
				invalidOk++
			})
		}
	})

	pct := func(a, b int) string { return fmt.Sprintf("%.1f%%", 100*float64(a)/float64(b)) }
	fmt.Fprintf(os.Stderr,
		"\n[strictness] Go baseline @ zig %s\n"+
			"[strictness]   valid-accepted   %d/%d (%s)\n"+
			"[strictness]   invalid-rejected %d/%d (%s)\n",
		corpus.Source.Zig,
		validOk, len(valid), pct(validOk, len(valid)),
		invalidOk, len(invalid), pct(invalidOk, len(invalid)))
}
